### PR TITLE
Pydantic v2

### LINKED ISSUE #102: Pydantic V2
Hi,

Thanks for this useful stuff. Could you update your repo with the latest pydantic V2? You may find useful migration tool here https://docs.pydantic.dev/latest/migration/ .

`File "/jupyterhub/4.0/etc/jupyterhub/jupyterhub_config_clean_virtenv_moss_test.py", line 8, in <module>
        import jupyterhub_moss  # noqa

      File "/jupyterhub/4.0/lib/python3.8/site-packages/jupyterhub_moss/__init__.py", line 3, in <module>
        from .spawner import MOSlurmSpawner

      File "/jupyterhub/4.0/lib/python3.8/site-packages/jupyterhub_moss/spawner.py", line 15, in <module>
        from .models import (

      File "/jupyterhub/4.0/lib/python3.8/site-packages/jupyterhub_moss/models.py", line 6, in <module>
        from pydantic import (

      File "/jupyterhub/4.0/lib/python3.8/site-packages/pydantic/__init__.py", line 206, in __getattr__
        return _getattr_migration(attr_name)

      File "/jupyterhub/4.0/lib/python3.8/site-packages/pydantic/_migration.py", line 285, in wrapper
        raise PydanticImportError(f'`{import_path}` has been removed in V2.')

    pydantic.errors.PydanticImportError: `pydantic:ConstrainedStr` has been removed in V2.

    For further information visit https://errors.pydantic.dev/2.0.2/u/import-error`

### PR DESCRIPTION
**Pull request by t20100**
This PR updates the code to run with pydantic v2 which was recently released and breaks a few APIs.

closes #102

### CODE REVIEW

File: jupyterhub_moss/models.py [hunk 1]
```diff
@@ -4,32 +4,29 @@
...
 
 
-class NonEmptyStr(ConstrainedStr):
-    min_length = 1
-    strip_whitespace = True
-
-
-def check_match_gpu(v: Optional[int], values: dict) -> Optional[int]:
-    if v is not None and v > 0 and values.get("gpu") == "":
+def check_match_gpu(v: Optional[int], info: FieldValidationInfo) -> Optional[int]:
```
Comment (t20100) [inline jupyterhub_moss/models.py:20]: validator function signature changed

File: jupyterhub_moss/models.py [hunk 2]
```diff
@@ -4,32 +4,29 @@
...
+    if v is not None and v > 0 and info.data.get("gpu") == "":
         return 0  # GPU explicitly disabled
     return v
 
 
 # models
 
 
-class PartitionResources(BaseModel, allow_mutation=False, extra=Extra.allow):
+class PartitionResources(BaseModel, frozen=True, extra="allow"):
```
Comment (t20100) [inline jupyterhub_moss/models.py:29]: config param names changed and use of `Extra....` is deprecated

File: jupyterhub_moss/models.py [hunk 3]
```diff
@@ -44,12 +41,10 @@ class PartitionResources(BaseModel, allow_mutation=False, extra=Extra.allow):
     max_runtime: PositiveInt
 
     # validators
-    _check_match_gpu = validator("max_ngpus", allow_reuse=True)(check_match_gpu)
+    _check_match_gpu = field_validator("max_ngpus")(check_match_gpu)
```
Comment (t20100) [inline jupyterhub_moss/models.py:44]: Name change and now useless `allow_reuse` param

File: jupyterhub_moss/models.py [hunk 4]
```diff
@@ -62,41 +57,40 @@ class PartitionAllResources(
...
         return v
 
 
-class PartitionConfig(BaseModel, allow_mutation=False, extra=Extra.forbid):
+class PartitionConfig(BaseModel, frozen=True, extra="forbid"):
     """Information about partition description and available environments"""
 
-    architecture = ""
-    description = ""
+    architecture: str = ""
```
Comment (t20100) [inline jupyterhub_moss/models.py:81]: Types looks to be needed for all field even with a default value

File: jupyterhub_moss/models.py [hunk 5]
```diff
@@ -62,41 +57,40 @@ class PartitionAllResources(
...
 
 
-class PartitionInfo(
-    PartitionConfig, PartitionResources, allow_mutation=False, extra=Extra.allow
-):
+class PartitionInfo(PartitionConfig, PartitionResources):
     """Complete information about a partition: config and resources"""
 
-    pass
+    model_config = ConfigDict(frozen=True, extra="allow")
```
Comment (t20100) [inline jupyterhub_moss/models.py:90]: Former way of setting this no longer work with multiple inheritance

File: jupyterhub_moss/models.py [hunk 6]
```diff
@@ -105,53 +99,59 @@ class _PartitionTraits(PartitionConfig, allow_mutation=False, extra=Extra.forbid
...
     """Configuration passed as ``partitions`` trait"""
 
-    __root__: Dict[str, _PartitionTraits]
+    root: Dict[str, _PartitionTraits]
+
+    model_config = ConfigDict(frozen=True, extra="forbid")
 
-    def dict(self, *args, **kwargs):
-        return {k: v.dict(*args, **kwargs) for k, v in self.__root__.items()}
+    def model_dump(self, *args, **kwargs):
```
Comment (t20100) [inline jupyterhub_moss/models.py:124]: `dict()` renamed `model_dump()`

File: jupyterhub_moss/models.py [hunk 7]
```diff
@@ -189,24 +189,22 @@ def has_no_newline(cls, v: str) -> str:
...
             parse_timelimit(v)  # Raises exception if malformed
         return v
 
-    _MEM_REGEXP = re.compile("^[0-9]*([0-9]+[KMGT])?$")
-
-    @validator("memory")
+    @field_validator("memory")
     def check_memory(cls, v: str) -> str:
-        if v and cls._MEM_REGEXP.match(v) is None:
+        if v and _MEM_REGEXP.match(v) is None:
```
Comment (t20100) [inline jupyterhub_moss/models.py:208]: There was trouble with access to class attribute

**Review by loichuder — APPROVED** [review]
Thanks for the comprehensive PR :heart:

**Pull request merged**